README.md
# Grid UV Packer

![grid-uv-packer logo](./img/grid-uv-packer-logo.png)

A pure-Python UV packer that packs irregularly shaped UV islands efficiently.

![example](./img/example.png)

grid-uv-packer is released with [GPL license](./COPYING.txt), same as
[Blender](https://www.blender.org/about/license/)

## Features

- Packs UV islands trying to utilize all available space.
- Uses multi-processing to run multiple iterations in parallel.

## Installation

- Download the source zip from [releases
  page](https://github.com/muhuk/grid-uv-packer/releases).  You need to click
  to expand `Assets` under the correct version and download the file named
  `guvp-#.#.zip` (where #.# is the version number)
- In Blender, open `Preferences` editor (`Edit -> Preferences`
  from the menu).
- Click `Install...` button and choose the zip file you have downloaded.
- Once the addon is installed, you **must** still enable it manually in
  `Preferences` view.  It is not automatically enabled.
- After installation, you can delete the zip file you have downloaded.

## Usage

Once you enable the addon, in UV editor a `Grid UV Pack` command will be
available under `UV` menu.  This command will create a modal dialog where you
can adjust operator parameters.  Operator will start to run once you click
`OK` on this modal dialog.

Note that operator only packs selected islands, same behavior as `Pack Islands`.

### Parameters

*Pack to*
: This is same as `Pack Islands` operator.  `Closest UDIM` packs into the tile
  closest to the selection's center, `Active UDIM` packs into the active tile
  (if a tiled image is loaded) or the tile closest to 2D cursor.

*Grid Size*
: Resolution of the grid used by the algorithm.  Increasing this parameter can
  result in more tightly packed UV maps at the cost of time.  This parameter
  should be adjusted with 2 things in mind; the number of islands and image
  dimensions.  As a rough estimate `64` and `128` is good for 1K images, `512`
  should only be used for 8K textures for very dense meshes.

*Rotate*
: Whether to rotate islands or not during packing.  This is same as `Pack Islands`.

*Margin*
: Spacing (in terms of UV space) in between islands.  Althogh `Pack Islands`
  also have a parameter with the same name, it works differently in this
  operator.  Margin is taken into account, but the result might have more spacing.

*Max Iterations*
: Maximum number of iterations.  When the calculation ends, result with the
  best coverage will be used.  Due to brute force nature of this operator,
  setting this to a low number will yield poor results, expecially with UV
  maps with many islands.  Note that calculation may be stopped before this
  number of iterations are run.

*Max Runtime*
: Maximum amount of time the calculation is allowed to run.  Set this to `0`
  to disable time limit.

*Random Seed*
: Seed used to initialize the random number generator.  If this is set to `0`
  a random number will be generated every time you run the operator.  You
  don't need to modify this normally.


CHANGELOG.md
# Changelog

## [Unreleased]

## [1.0] - 2023-02-14

### Added

- UDIM support is added.  `Pact to` options `Active UDIM` and `Closest UDIM`
  should work same as `py.ops.uv.pack_islands()`.
- You can interrupt calculation via `ESC` key.  Note that this does not cancel
  the operation.  If a solution with better coverage is already found, UVs
  will be updated.  (You can undo the operator)

### Changed

- Moved from `Unwrap` menu to `UV`
  menu. [b659c64](https://github.com/muhuk/grid-uv-packer/commit/37f5e510f557d3d6eb7c1956eb2515575b659c64)
- When `Max Iterations` is set to `1` packer is run single threaded.  Note
  that you need a lot more than one iteration to get decent results, this is
  intended for
  debugging. [8f36daf](https://github.com/muhuk/grid-uv-packer/commit/06f16e6ce46babb7420e6b2053cf9bc038f36daf)
- Previously any result with 20% coverage or better was considered successful
  and written to UVs.  Now only the results with better coverage than input
  UVs are written. [fc949cb](https://github.com/muhuk/grid-uv-packer/commit/df5e0f9bc89937fb120cd71926dedbea7fc949cb)

### Fixed

- Fixed an issue where rotated islands were overlapping with other
  islands. [6ae83fe](https://github.com/muhuk/grid-uv-packer/commit/f17f337423dbe9ffad4e4641b6be9a5ab6ae83fe)
- Earlier packing tended to produce results that cover a rectangular area
  (mostly x axis, y axis underutilized) instead of the whole UV square.  This
  is fixed with
  [e107900](https://github.com/muhuk/grid-uv-packer/commit/e85f809bc52e5fc591e191cc990ccaf6ee107900)
  and
  [3863a48](https://github.com/muhuk/grid-uv-packer/commit/4ce97666e49fcaabe248a31e00beb54073863a48).

### Removed

- Removed baseline fitness calculation.  This was not something users were
  seeing but before UDIM support, if all islands were not in UV unit square
  the operator gave an error.  Otherwise a baseline fitness was being
  calculated and the UVs were only updated if the grid packer's result was
  better than the original UV map.  When the UDIM support added calculating
  the baseline fitness became difficult so it was removed.  Now the UVs are
  updated provided that all islands are placed on the grid (a solution is
  found).  If the result is not good, you can undo.

## [0.3] - 2022-12-29

### Added

- Added setting for maximum number of iterations.
- Added setting for maximum run time.  Set this to `0` to disable run time
  limit.
- Implemented island rotation.  This setting is on by default, you can
  distable it by unchecking `rotate`.

### Changed

- Minimum required version of Blender is now `3.4.0`.

## [0.2] - 2022-12-15

### Added
- Made random seed configurable.  Set seed to `0` to get a random
  seed. ([57d1536](https://github.com/muhuk/grid-uv-packer/commit/e17443220fb5f74daaffddead07c389f657d1536))

### Changed
- When UV islands are out of bounds `grid_pack` command will be cancelled
  instead of throwing an
  exception. ([0d2dc3a](https://github.com/muhuk/grid-uv-packer/commit/ecb3ea902b6cd78ffea7b746060d5dd230d2dc3a))
- Improved coverage (fitness)
  ([68adb5c](https://github.com/muhuk/grid-uv-packer/commit/49f6bd25592c167362d19a2952509038a68adb5c))
  calculation & margin calculation
  ([4dc76e8](https://github.com/muhuk/grid-uv-packer/commit/d69f0015395c33149028cadb80e448eca4dc76e8)).

### Fixed
- Fixed the non-breaking exception when parsing
  `bl_info`. ([7e1fcd9](https://github.com/muhuk/grid-uv-packer/commit/214566e9cac4113241374e20ba3631dc37e1fcd9))

## [0.1] - 2022-12-13

### Added
- Initial proof of concept release.  Not ready for production use.

[unreleased]: https://github.com/muhuk/grid-uv-packer/compare/v1.0...HEAD
[1.0]: https://github.com/muhuk/grid-uv-packer/compare/v0.3...v1.0
[0.3]: https://github.com/muhuk/grid-uv-packer/compare/v0.2...v0.3
[0.2]: https://github.com/muhuk/grid-uv-packer/compare/v0.1...v0.2
[0.1]: https://github.com/muhuk/grid-uv-packer/releases/tag/v0.1


guvp/tests.py
import contextlib
import unittest

import bpy
import bmesh
from mathutils import Vector
from typing import Optional

from guvp import data


class IslandTest(unittest.TestCase):
    def assertEqualVector(
            self,
            first: Vector,
            second: Vector,
            msg: Optional[str] = None
    ) -> None:
        return self.assertAlmostEqual((first - second).length, 0.0, msg=msg)

    def test_island_creation_from_bmesh_faces(self):
        bpy.ops.mesh.primitive_plane_add()
        obj = bpy.context.selected_objects[0]
        bm = bmesh.new()
        bm.from_mesh(obj.data)
        island = data.Island.from_faces(bm, {0})
        # Check that size is correct
        self.assertAlmostEqual(island.size[0], 1.0)
        self.assertAlmostEqual(island.size[1], 1.0)
        # Check that UVs are set right.
        self.assertEqual(len(island.uvs), 4)
        self.assertEqualVector(island.uvs[0], Vector((0.0, 0.0)))
        self.assertEqualVector(island.uvs[1], Vector((1.0, 0.0)))
        self.assertEqualVector(island.uvs[2], Vector((1.0, 1.0)))
        self.assertEqualVector(island.uvs[3], Vector((0.0, 1.0)))
        # bm.to_mesh(obj.data)
        bm.free()
        del(bm)
        bpy.ops.object.delete()
        del(obj)


def run():
    assert(bpy.context.mode == 'OBJECT')
    suite = unittest.defaultTestLoader.loadTestsFromTestCase(IslandTest)
    runner = unittest.TextTestRunner()
    with saved_selection():
        result = runner.run(suite)
    return result


@contextlib.contextmanager
def saved_selection():
    # Save selection
    saved_selection = bpy.context.selected_objects
    saved_active = bpy.context.active_object
    bpy.ops.object.select_all(action='DESELECT')
    # Run suite
    try:
        yield
    finally:
        # Restore selection
        for obj in saved_selection:
            obj.select_set(True)
        if saved_active is not None:
            bpy.context.view_layer.objects.active = saved_active


guvp/data.py
from __future__ import annotations
from dataclasses import dataclass
import enum
import math
from typing import (
    Any,
    Dict,
    Iterator,
    List,
    NamedTuple,
    Set,
    Type
)

import bmesh                  # type: ignore
from mathutils import Vector  # type: ignore
import mathutils.geometry     # type: ignore

from guvp import discrete


CollisionResult = enum.Enum('CollisionResult', 'NO YES OUT_OF_BOUNDS')

FaceUVs = Dict[int, Vector]

# face id -> (face loop id -> UV)
IslandUVs = Dict[int, FaceUVs]


class GridPacker:
    def __init__(self, initial_size: int, islands: List[Island]) -> None:
        self._utilized_area = discrete.Size.zero()
        self._mask = discrete.Grid.empty(width=initial_size, height=initial_size)
        self._islands: List[IslandPlacement] = [
            IslandPlacement(offset=discrete.CellCoord.zero(), island=island)
            for island in islands
        ]

    @property
    def fitness(self) -> float:
        # TODO: Calculate actual fitness.
        #       Return value between 0.0 - 1.0
        return 0.0

    def run(self) -> None:
        filled_x: int = 0
        for ip in self._islands:
            # using current offset try collision: `mask & island`
            # if collision:
            #    increment offset
            # else:
            #    give new offset to island
            #    update the mask (stamp island)
            #    update utilized area
            pass
        for ip in self._islands:
            ip.offset = discrete.CellCoord(filled_x, 0)
            filled_x += ip.island.mask.size.width
            (uw, uh) = self._utilized_area
            self.utilized_area = discrete.Size(
                width=filled_x,
                height=max(ip.island.mask.size.height, uh)
            )
        print("Utilized area: {0}".format(self.utilized_area))

    def write(self, bm: bmesh.types.BMesh) -> None:
        for ip in self._islands:
            ip.write_uvs(bm)

    def _check_collision(self, ip: IslandPlacement) -> CollisionResult:
        if ip.offset.x + ip.island.mask.size.width > self._mask.size.width:
            return CollisionResult.OUT_OF_BOUNDS
        if ip.offset.y + ip.island.mask.size.height > self._mask.size.height:
            return CollisionResult.OUT_OF_BOUNDS
        raise NotImplementedError()


@dataclass(frozen=True)
class Island:
    face_ids: set[int]
    uvs: IslandUVs
    cell_size: float
    mask: Grid

    @classmethod
    def from_faces(
            cls: Type[Island],
            bm: bmesh.types.BMesh,
            face_ids: set[int],
            cell_size: float
            # TODO: add margin parameters
    ) -> Island:
        (uvs, size, offset) = cls._calculate_uvs_and_size(bm, face_ids)
        mask = discrete.Grid.empty(
            width=math.ceil(size.x / cell_size),
            height=math.ceil(size.y / cell_size)
        )
        cls._fill_mask(bm, face_ids, offset, cell_size, mask)
        print(mask)
        mask.draw_str()
        return cls(
            face_ids=face_ids,
            uvs=uvs,
            cell_size=cell_size,
            mask=mask
        )

    def write_uvs(self, bm: bmesh.types.BMesh, offset: discrete.CellCoord) -> None:
        uv_ident = bm.loops.layers.uv.verify()
        offset_vec: Vector = Vector(offset) * self.cell_size
        for face_id in self.face_ids:
            for face_loop in bm.faces[face_id].loops:
                assert(face_loop.index in self.uvs[face_id])
                face_loop[uv_ident].uv = self.uvs[face_id][face_loop.index]
                face_loop[uv_ident].uv += offset_vec

    @staticmethod
    def _calculate_uvs_and_size(
            bm: bmesh.types.BMesh,
            face_ids: set[int]
    ) -> tuple[IslandUVs, Vector, Vector]:
        uvs: IslandUVs = {}
        uv_ident = bm.loops.layers.uv.verify()
        (u_min, v_min) = (math.inf, math.inf)
        (u_max, v_max) = (-math.inf, -math.inf)
        for face_id in face_ids:
            face_uvs: FaceUVs = {}
            for face_loop in bm.faces[face_id].loops:
                # Store UV
                face_uvs[face_loop.index] = face_loop[uv_ident].uv
                # Update UV bounds
                (u, v) = face_loop[uv_ident].uv
                u_min = min(u, u_min)
                v_min = min(v, v_min)
                u_max = max(u, u_max)
                v_max = max(v, v_max)
                del u, v
            uvs[face_id] = face_uvs
        size = Vector((u_max - u_min, v_max - v_min))
        # Offset everything by (-u_min, -v_min)
        offset = Vector((u_min, v_min))
        uvs = {k: {loop_id: uv - offset for loop_id, uv in fuv.items()}
               for k, fuv in uvs.items()}
        return (uvs, size, offset)

    @staticmethod
    def _fill_mask(
            bm: bmesh.types.BMesh,
            face_ids: set[int],
            offset: Vector,
            cell_size: float,
            mask: discrete.Grid
    ):
        open_cells: Set[discrete.CellCoord] = set(mask)
        uv_ident = bm.loops.layers.uv.verify()

        for face_id in face_ids:
            loop_uvs = [face_loop[uv_ident].uv
                        for face_loop in bm.faces[face_id].loops]
            # When we use triangulate, we are assuming the face,
            # when it is an n-gon, is convex.  There is basically no way
            # to triangulate a concave n-gon without triangulating it first.
            for face_tri in Triangle2D.triangulate(
                [(u, v) for (u, v) in loop_uvs]
            ):
                hit_cells: Set[discrete.CellCoord] = set()
                for open_cell_id in open_cells:
                    # We need to invert y-axis because UVs
                    # use vertically increasing y-axis.
                    cell_x: int = open_cell_id[0]
                    cell_y: int = mask.size.height - open_cell_id[1]
                    # triangulate the cell's quad.
                    x = float(offset.x + cell_x * cell_size)
                    y = float(offset.y + cell_y * cell_size)
                    grid_tris = Triangle2D.triangulate([
                        Vector((x, y)),
                        Vector((x, y + cell_size)),
                        Vector((x + cell_size, y + cell_size)),
                        Vector((x + cell_size, y))
                    ])
                    for grid_tri in grid_tris:
                        if grid_tri.intersect(face_tri):
                            hit_cells.add(open_cell_id)
                            mask[open_cell_id] = True
                            break
                open_cells -= hit_cells


@dataclass
class IslandPlacement:
    offset: discrete.CellCoord
    # rotation: Enum
    #
    #   see: https://docs.python.org/3/library/enum.html
    island: Island

    def write_uvs(self, bm: bmesh.types.BMesh) -> None:
        self.island.write_uvs(bm, self.offset)


class Triangle2D:
    def __init__(self, a: Vector, b: Vector, c: Vector):
        self.a = a
        self.b = b
        self.c = c

    def intersect(self, other: Triangle2D) -> bool:
        return mathutils.geometry.intersect_tri_tri_2d(
            self.a,
            self.b,
            self.c,
            other.a,
            other.b,
            other.c
        )

    @classmethod
    def triangulate(cls, verts: List[Vector]) -> List[Triangle2D]:
        n: int = len(verts)
        vert_coords = list(map(Vector, verts))
        edges = [(a, (a + 1) % n) for a in range(n)]
        faces = [list(range(n))]
        output_type: int = 0  # triangles with convex hull
        epsilon: float = 1.0 / 1_000_000.0
        delanuay_result = mathutils.geometry.delaunay_2d_cdt(
            vert_coords,
            edges,
            faces,
            output_type,
            epsilon
        )
        (t_verts, _, t_faces, _, _, _) = delanuay_result
        result: List[Triangle2D] = []
        for vert_ids in t_faces:
            assert(len(vert_ids) == 3)
            verts = [t_verts[vert_id] for vert_id in vert_ids]
            result.append(cls(verts[0], verts[1], verts[2]))
        return result


guvp/debug.py
# <pep8-80 compliant>

# grid-uv-packer packs irregularly shaped UV islands efficiently.
# Copyright (C) 2022  Atamert Ölçgen
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.


import bpy

from guvp import constants


def is_debug() -> bool:
    addon_prefs = \
        bpy.context.preferences.addons[constants.ADDON_NAME].preferences
    return addon_prefs.debug_mode


def print_(fmt_str: str, *args, **kwargs) -> None:
    if is_debug():
        print(fmt_str.format(*args, **kwargs))


guvp/props.py
# <pep8-80 compliant>

# grid-uv-packer packs irregularly shaped UV islands efficiently.
# Copyright (C) 2022  Atamert Ölçgen
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.


import bpy          # type: ignore

from guvp import constants


class GridUVPackerAddonPreferences(bpy.types.AddonPreferences):
    bl_idname = constants.ADDON_NAME

    debug_mode: bpy.props.BoolProperty(
        name="Debug mode",
        description="Enable debug output.",
        default=constants.DEBUG_MODE_DEFAULT
    )

    def draw(self, context):
        layout = self.layout
        col = layout.column(align=True, heading="Developer Preferences")
        col.prop(self, "debug_mode")


guvp/constants.py
# <pep8-80 compliant>

# grid-uv-packer packs irregularly shaped UV islands efficiently.
# Copyright (C) 2022  Atamert Ölçgen
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.


import enum

ADDON_NAME = __package__

DEBUG_MODE_DEFAULT: bool = False

# Grow if utilized area is larger than this.
GROW_AREA_RATIO = 0.85

MAX_GROW_COUNT = 2

MAX_ITERATIONS_DEFAULT = 500
MAX_ITERATIONS_LIMIT = 10000

# Hard limit for tries
MAX_PLACEMENT_RETRIES = 100_000

MAX_RUNTIME_DEFAULT = 90
MAX_RUNTIME_LIMIT = 3600  # 1 hour

OPERATOR_TIMER_TIME_STEP = 0.5  # seconds

SEARCH_START_RESET_CHANCE = 0.05

SEED_MAX = 2 ** 31 - 1

# Rotation is CW, like in Blender.
Rotation = enum.Enum(
    'Rotation',
    {'NONE': 0.0,
     'DEGREES_90': 90.0,
     'DEGREES_180': 180.0,
     'DEGREES_270': 270.0})
ALL_ROTATIONS = (
    Rotation.NONE,
    Rotation.DEGREES_90,
    Rotation.DEGREES_180,
    Rotation.DEGREES_270
)


guvp/__init__.py
# <pep8-80 compliant>

# grid-uv-packer packs irregularly shaped UV islands efficiently.
# Copyright (C) 2022  Atamert Ölçgen
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.


from __future__ import annotations

if "bpy" in locals():
    import importlib
    for mod in [                  # noqa: F821
            constants,            # noqa: F821
            continuous,           # noqa: F821
            debug,                # noqa: F821
            discrete,             # noqa: F821
            packing,              # noqa: F821
            props                 # noqa: F821
    ]:
        print("reloading {0}".format(mod))
        importlib.reload(mod)
else:
    # stdlib
    from functools import reduce
    import math
    import random
    import time
    from typing import (Iterable, List, Optional, Set)
    # blender
    import bpy                                                  # type: ignore
    from bpy_extras.bmesh_utils import bmesh_linked_uv_islands  # type: ignore
    import bmesh                                                # type: ignore
    from mathutils import Vector                                # type: ignore
    # addon
    from guvp import (                                          # noqa: F401
        constants,
        continuous,
        debug,
        discrete,
        packing,
        props
    )


bl_info = {
    "name": "Grid UV Packer",
    "description": "A pure-Python UV packer.",
    "author": "Atamert Ölçgen",
    "version": (1, 0),
    "blender": (3, 4, 0),
    "location": "TBD",
    "tracker_url": "https://github.com/muhuk/grid_uv_packer",
    "support": "COMMUNITY",
    "category": "UV"
}


class GridUVPackOperator(bpy.types.Operator):
    """Grid UV Pack Operator"""
    bl_idname = "uv.grid_pack"
    bl_label = "Grid UV Pack"
    bl_options = {'UNDO'}

    pack_to: bpy.props.EnumProperty(    # type: ignore
        name="Pack to",
        default="closest",              # noqa: F821
        items=(
            ('closest', "Closest UDIM", "", 'NONE', 1),  # noqa: F821
            ('active', "Active UDIM", "", 'NONE', 2),    # noqa: F821
        )
    )
    grid_size: bpy.props.EnumProperty(  # type: ignore
        name="Grid Size",               # noqa: F722
        default="128",
        items=(
            ('64', "64", "", 'NONE', 64),     # noqa: F821
            ('128', "128", "", 'NONE', 128),  # noqa: F821
            ('256', "256", "", 'NONE', 256),  # noqa: F821
            ('512', "512", "", 'NONE', 512)   # noqa: F821
        ),
        options={'SKIP_SAVE'}                 # noqa: F821
    )
    rotate: bpy.props.BoolProperty(           # type: ignore
        name="Rotate",                        # noqa: F821
        description="Rotate islands for best fit.",
        default=True,
        options={'SKIP_SAVE'}                 # noqa: F821
    )
    margin: bpy.props.FloatProperty(          # type: ignore
        name="Margin",                        # noqa: F821
        description="Space between islands.",
        default=0.01,
        min=0.0,
        max=1.0,
        precision=3,
        options={'SKIP_SAVE'}                 # noqa: F821
    )
    max_iterations: bpy.props.IntProperty(    # type: ignore
        name="Max Iterations",
        description="Maximum number of iterations.",
        default=constants.MAX_ITERATIONS_DEFAULT,
        min=1,
        max=constants.MAX_ITERATIONS_LIMIT,
        options={'SKIP_SAVE'}                 # noqa: F821
    )
    max_runtime: bpy.props.IntProperty(       # type: ignore
        name="Max Runtime",
        description="Maximum time the calculation can take, in seconds.",
        default=constants.MAX_RUNTIME_DEFAULT,
        min=0,
        max=constants.MAX_RUNTIME_LIMIT,
        options={'SKIP_SAVE'}                 # noqa: F821
    )
    seed: bpy.props.IntProperty(              # type: ignore
        name="Random Seed",
        description="Seed of the random generator.",
        default=0,
        min=0,
        max=constants.SEED_MAX,
        options={'SKIP_SAVE'}                 # noqa: F821
    )

    def __init__(self):
        # Size of one grid cell (square) in UV coordinate system.
        self.cell_size: float
        # This is the actual seed used in calculations.
        # If 'seed' is set to 0, 'random_seed' takes a random value.
        self.random_seed: int
        self.start_time_ns: int
        self.end_time_ns: Optional[int]
        self.bm: bmesh.types.BMesh
        self.island_face_ids: List[set[int]]
        self.udim_offset: Vector
        self.packer: packing.GridPacker
        self._timer: bpy.types.Timer

    @classmethod
    def poll(cls, context):
        return context.active_object is not None and \
            context.active_object.type == 'MESH' and \
            context.active_object.data.uv_layers.active is not None and \
            context.mode == 'EDIT_MESH'

    def execute(self, context: bpy.types.Context) -> Set[str]:
        assert context.active_object is not None
        assert context.active_object.type == 'MESH'
        assert context.active_object.data.uv_layers.active is not None
        assert context.mode == 'EDIT_MESH'
        assert 0.0 <= self.margin <= 1.0

        self.cell_size = 1.0 / int(self.grid_size)
        self.random_seed = self.seed if self.seed != 0 \
            else random.randint(0, constants.SEED_MAX)
        self.start_time_ns = time.time_ns()

        self.bm = bmesh.from_edit_mesh(context.active_object.data)
        self.bm.verts.ensure_lookup_table()
        self.bm.edges.ensure_lookup_table()
        self.bm.faces.ensure_lookup_table()

        self.island_face_ids = self._island_face_ids(self.bm)

        # Calculate UDIM offset
        self._calculate_udim_offset(
            context,
            reduce(lambda a, b: set(a) | b, self.island_face_ids, set())
        )
        debug.print_(
            "Packing to {} UDIM tile.  Offset is {!r}",
            self.pack_to,
            self.udim_offset
        )

        debug.print_("Seed being used is: {}", self.random_seed)
        if self.max_iterations == 1:
            return self.execute_single(context)
        else:
            return self.execute_parallel(context)

    def execute_single(self, context: bpy.types.Context) -> Set[str]:
        debug.print_("Running a single iteration.")

        self.packer = packing.GridPackerSingle(
            initial_size=int(self.grid_size),
            islands=[
                continuous.Island.from_faces(
                    self.bm,
                    face_ids,
                    self.cell_size,
                    self.margin
                )
                for face_ids in self.island_face_ids
            ],
            rotate=self.rotate,
            random_seed=self.random_seed
        )
        self.packer.run_single()

        return self.finish(context)

    def execute_parallel(self, context: bpy.types.Context) -> Set[str]:
        if self.max_runtime > 0:
            self.end_time_ns = self.start_time_ns + \
                self.max_runtime * 1_000_000_000
        else:
            self.end_time_ns = None

        wm: bpy.types.WindowManager = context.window_manager
        wm.progress_begin(0, 10000)
        wm.progress_update(1)

        self.packer = packing.GridPackerParallel(
            initial_size=int(self.grid_size),
            islands=[
                continuous.Island.from_faces(
                    self.bm,
                    face_ids,
                    self.cell_size,
                    self.margin
                )
                for face_ids in self.island_face_ids
            ],
            rotate=self.rotate,
            random_seed=self.random_seed
        )
        self.packer.run()

        self._timer = wm.event_timer_add(
            constants.OPERATOR_TIMER_TIME_STEP,
            window=context.window
        )
        context.window_manager.modal_handler_add(self)
        return {'RUNNING_MODAL'}

    def modal(self,
              context: bpy.types.Context,
              event: bpy.types.Event) -> Set[str]:
        assert type(self.packer) is packing.GridPackerParallel
        wm: bpy.types.WindowManager = context.window_manager
        should_continue = True
        if event.type == 'TIMER':
            if self.packer.iterations_completed < self.max_iterations \
               and (self.end_time_ns is None
                    or time.time_ns() < self.end_time_ns):
                wm.progress_update(int(
                    float(self.packer.iterations_completed)
                    / self.max_iterations * 10000
                ))
                debug.print_(
                    "Iterations so far {}, fitness {:.2f}%",
                    self.packer.iterations_completed,
                    self.packer.fitness * 100.0
                )
            else:
                should_continue = False
        elif event.type == 'ESC':
            should_continue = False

        if should_continue:
            return {'RUNNING_MODAL'}
        else:
            debug.print_("Stopping packer.")
            self.packer.stop()
            wm.progress_end()
            wm.event_timer_remove(self._timer)
            return self.finish(context)

    def finish(self, context: bpy.types.Context) -> Set[str]:
        debug.print_(
            "Fitness is {:.2f}%, scaling factor is {:.3f}",
            self.packer.fitness * 100,
            self.packer.scaling_factor
        )
        if self.packer.fitness > 0.00001:
            self.packer.write(self.bm, self.udim_offset)
            bmesh.update_edit_mesh(
                mesh=context.active_object.data,
                loop_triangles=False,
                destructive=False
            )
            self.report({'INFO'}, "UVs updated.")
        else:
            self.report({'INFO'}, "UVs are not changed.")

        debug.print_(
            "Total time: {:.3f}",
            (time.time_ns() - self.start_time_ns) / 1_000_000_000
        )
        return {'FINISHED'}

    def invoke(self, context, event) -> Set[str]:
        wm: bpy.types.WindowManager = context.window_manager
        return wm.invoke_props_dialog(self)

    def _calculate_udim_offset(
            self,
            context: bpy.types.Context,
            face_ids: Iterable[int]
    ) -> None:
        if self.pack_to == 'active':
            self.udim_offset = self._calculate_udim_offset_active(context)
        elif self.pack_to == 'closest':
            center: Vector = self._calculate_center(
                self.bm,
                face_ids
            )
            self.udim_offset = self._calculate_udim_offset_closest(center)

    @staticmethod
    def _calculate_center(
            bm: bmesh.types.BMesh,
            face_ids: Iterable[int]
    ) -> Vector:
        uv_ident = bm.loops.layers.uv.active

        u_min: float = math.inf
        u_max: float = -math.inf
        v_min: float = math.inf
        v_max: float = -math.inf
        for face_id in face_ids:
            for loop in bm.faces[face_id].loops:
                (u, v) = loop[uv_ident].uv
                u_min = min(u_min, u)
                u_max = max(u_max, u)
                v_min = min(v_min, v)
                v_max = max(v_max, v)
        return Vector(((u_min + u_max) * 0.5, (v_min + v_max) * 0.5))

    @staticmethod
    def _island_face_ids(bm: bpy.types.BMesh) -> List[set[int]]:
        """Calculate sets of faces that form a UV island."""
        uv_ident = bm.loops.layers.uv.active
        island_faces = bmesh_linked_uv_islands(bm, uv_ident)
        result = []
        island_selected: bool
        for faces in island_faces:
            island_selected = True
            for face in faces:
                for loop in face.loops:
                    if loop[uv_ident].select is False:
                        island_selected = False
                        break  # if one face-corner is unselected
                if island_selected is False:
                    break  # if one of the faces in island is unselected
            if island_selected:
                result.append(set([f.index for f in faces]))
        return result

    @staticmethod
    def _calculate_udim_offset_active(context: bpy.types.Context) -> Vector:
        # Active UDIM image tile or, if no image is available, the UDIM grid
        # tile where the 2D cursor is located.
        tile_x: int
        tile_y: int
        if context.space_data.image is not None:
            img: bpy.types.Image = context.space_data.image
            tile_x = img.tiles.active.number % 10
            if tile_x == 0:
                tile_x = 10
            tile_x -= 1
            tile_y = img.tiles.active.number // 10
            tile_y -= 100
        else:
            tile_x = math.floor(context.space_data.cursor_location[0])
            tile_y = math.floor(context.space_data.cursor_location[1])
        return Vector((tile_x, tile_y))

    @staticmethod
    def _calculate_udim_offset_closest(center: Vector) -> Vector:
        tile_x: int = math.floor(center.x)
        tile_y: int = math.floor(center.y)
        return Vector((tile_x, tile_y))


def menu_draw(self, _context):
    self.layout.separator()
    self.layout.operator_context = 'INVOKE_DEFAULT'
    self.layout.operator("uv.grid_pack")


def register():
    # Register props
    bpy.utils.register_class(props.GridUVPackerAddonPreferences)

    # Register operations
    bpy.utils.register_class(GridUVPackOperator)

    # Register UI
    bpy.types.IMAGE_MT_uvs.append(menu_draw)


def unregister():
    # Unregister UI
    bpy.types.IMAGE_MT_uvs.remove(menu_draw)

    # Unregister operations
    bpy.utils.unregister_class(GridUVPackOperator)

    # Unregister props
    bpy.utils.unregister_class(props.GridUVPackerAddonPreferences)


if __name__ == "__main__":
    register()

    # test call
    bpy.ops.uv.grid_pack()
